ModMapLayer positioning

Jadams505 opened the pull request:
### What is the new feature?
Allows modders to position their `ModMapLayers` between vanilla layers. (Feature 1 of #4583)
It adds the following methods to `ModMapLayer`:

- `public virtual Position GetDefaultPosition()`
- `public virtual IEnumerable<Position> GetModdedConstraints()`

### Why should this be part of tModLoader?
Currently map layers are added after all vanilla layers. This means that modded layers are drawn on top of vanilla layers. By being able to position modded layers it will be easier to deal with cases where icons overlap.
### Are there alternative designs?
The positioning approach used is the same as was used for `ExtraJumps` (#3552), but there are also the designs of `PlayerDrawLayers`, and `BuilderToggles`.
### Sample usage for the new feature
Indicates that the new layer will be ordered before the vanilla pings layer.
```c#
public override Position GetDefaultPosition() {
	return new Before(IMapLayer.Pings);
}
```
Specifies that this layer will show up in ordering after `ExampleMapLayer`. (assuming they are in the same vanilla slot)
```c#
public override IEnumerable<Position> GetModdedConstraints() {
	yield return new After(ModContent.GetInstance<ExampleMapLayer>());
}
```
### ExampleMod updates
Added `ExampleItemMapLayer` which makes any `ExampleItems` that are on the ground in the world show up on the map. It showcases how to use `GetModdedConstraints()` and orders itself before `ExampleMapLayer`.

### Porting Notes
Shouldn't be any breaking changes. Use the new methods if you want.

Jadams505 committed:
position implementation

Jadams505 committed:
change implementation to match ExtraJump

Jadams505 committed:
added ExampleItemMapLayer

 added a map layer to showcase GetModdedConstraints()

Jadams505 committed:
require vanilla layers in GetDefaultPosition()

Jadams505 committed:
ModMapLayer docs

Jadams505 committed:
format fixes

Chicken-Bones commented on ExampleMod/Common/ExampleMapLayer.cs:
```diff
@@ -7,12 +10,19 @@
 using Terraria.Map;
 using Terraria.ModLoader;
 using Terraria.UI;
+using static Terraria.Map.IMapLayer;
```
We don't normally support `using static` in `ExampleMod` because it can be a bit annoying to teach. I suppose it's needed here because `Position` is defined in an interface?

Jadams505 replied:
> I suppose it's needed here because `Position` is defined in an interface?

That is correct I opted to be explicit with `IMapLayer.Pings`, so that you could use autocomplete to choose the right layer, but I can remove using static and be explicit with `IMapLayer.Position` if it is confusing for new modders.

Actually I was thinking technically the positioning logic is never ran on vanilla layers which only implement `IMapLayer`, so technically we could just add the positioning hooks to `ModMapLayer` and leave `IMapLayer` unchanged. This would move `Position` into `ModMapLayer` and thus eliminate the need for `using static`. Thoughts?

Chicken-Bones replied:
That is the best solution, please do so

Jadams505 committed:
move Position to ModMapLayer

direwolf420 commented on patches/tModLoader/Terraria/ModLoader/ModMapLayer.cs:
```diff
@@ -9,6 +10,25 @@ public abstract class ModMapLayer : ModType, IMapLayer
+	/// Returns the map layer's default position in regard to vanilla's map layer ordering. Make use of e.g. <see cref="Before"/>/<see cref="After"/>, and provide a map layer.<br/><br/>
+	///
+	/// <b>NOTE:</b> The position must specify a vanilla <see cref="IMapLayer"/> otherwise an exception will be thrown.<br/>
+	/// By default, this hook positions this map layer after all vanilla layers.
+	/// </summary>
+	public virtual Position GetDefaultPosition() => new Before(null);
```
I find the syntax a bit strange. docs say `after all vanilla layers.`, but the code says `new Before(null)`, which implies it's before all vanilla layers. Is passing null and how it affects position documented to reflect this unintuitive relationship?

Jadams505 replied:
`Before(null)` means this layer should be before nothing, which means it is after everything. I don't love it either, but that's how `ExtraJump` does it and I figured they should be consistent.
I can add special positions `BeforeFirstVanillaLayer` and `AfterLastVanillaLayer` like `PlayerDrawLayers` to make it more clear.

direwolf420 commented on patches/tModLoader/Terraria/ModLoader/ModMapLayer.cs:
```diff
@@ -9,6 +10,25 @@ public abstract class ModMapLayer : ModType, IMapLayer
+
+	/// <summary>
+	/// Modded layers are placed between vanilla layers via <see cref="GetDefaultPosition"/> and, by default, are sorted in load order.<br/>
+	/// This hook allow you to sort this map layer before/after other modded layers that were placed between the same two vanilla layers.<br/>
```
allow -> allows

Jadams505 committed:
BeforeFirstVanillaLayer and AfterLastVanillaLayer

passing null to Before/After in GetDefaultPosition() will now throw due to its ambiguous meaning.

JavidPack committed:
Consistent style, selective map mode drawing example, documentation

JavidPack commented:
Thanks a ton, this was well done.
